auto_dev/protocols/formatter.py

```diff
@@ -0,0 +1,192 @@
+import textwrap
+
+from proto_schema_parser import ast
+from proto_schema_parser.ast import (
+    FileElement,
+    File,
+    Import,
+    Package,
+    Option,
+    Extension,
+    Service,
+    MessageElement,
+    Comment,
+    Field,
+    Group,
+    OneOf,
+    ExtensionRange,
+    Reserved,
+    Message,
+    Enum,
+    MapField,
+    MessageValue,
+    EnumElement,
+    FieldCardinality,
+)
+
+from auto_dev.protocols import adapters
+from auto_dev.protocols.adapters import FileAdapter, MessageAdapter
+from auto_dev.protocols.primitives import PRIMITIVE_TYPE_MAP
+
+
+def render_field(field: Field) -> str:
+    field_type = PRIMITIVE_TYPE_MAP.get(field.type, field.type)
+    match field.cardinality:
+        case FieldCardinality.REQUIRED | None:
+            return f"{field_type}"
+        case FieldCardinality.OPTIONAL:
+            return f"{field_type} | None"
+        case FieldCardinality.REPEATED:
+            return f"list[{field_type}]"
+        case _:
+            raise TypeError(f"Unexpected cardinality: {field.cardinality}")
+
+
+def render_attribute(element: MessageElement | MessageAdapter, prefix: str = "") -> str:
+    match type(element):
+        case ast.Comment:
+            return f"# {element.text}"
+        case ast.Field:
+            return f"{element.name}: {render_field(element)}"
+        case ast.OneOf:
+            if not all(isinstance(e, Field) for e in element.elements):
+                raise NotImplementedError("Only implemented OneOf for Field")
+            inner = " | ".join(render_field(e) for e in element.elements)
+            return f"{element.name}: {inner}"
+        case adapters.MessageAdapter:
+            elements = sorted(element.elements, key=lambda e: not isinstance(e, (MessageAdapter, ast.Enum)))
```

Comment:
sorting is important, because if fields are created first that reference nested types then python will throw.